### PR TITLE
fix: handle empty list with schema in table creation

### PR DESCRIPTION
**Pull request by tristanz**
## Summary
Fixes IndexError when creating tables with empty list data and a provided schema. Previously, `_into_pyarrow_reader()` would attempt to access `data[0]` on empty lists, causing an IndexError. Now properly handles empty lists by using the provided schema.

Also adds regression tests for GitHub issues #1968 and #303 to prevent future regressions with empty table scenarios.

## Changes
- Fix IndexError in `_into_pyarrow_reader()` for empty list + schema case
- Add Optional[pa.Schema] parameter to handle empty data gracefully  
- Add `test_create_table_empty_list_with_schema` for the IndexError fix
- Add `test_create_empty_then_add_data` for issue #1968
- Add `test_search_empty_table` for issue #303

## Test plan
- [x] All new regression tests pass
- [x] Existing tests continue to pass
- [x] Code formatted with `make format`

### CODE REVIEW

File: python/python/lancedb/table.py [hunk 1]
```diff
@@ -123,6 +125,11 @@ def _into_pyarrow_reader(data) -> pa.RecordBatchReader:
         raise ValueError("Cannot add a single dictionary to a table. Use a list.")
 
     if isinstance(data, list):
+        # Handle empty list case
+        if not data:
```
Comment (eddyxu) [inline python/python/lancedb/table.py:129]: Should we check `schema is not None` here?

Comment (tristanz) [reply python/python/lancedb/table.py:129]: Yes!

**Review by eddyxu — APPROVED** [review]
Thanks @tristanz !
